lib/sb3_schema.json

```diff
@@ -0,0 +1,368 @@
+{
+    "id": "https://scratch.mit.edu/sb3_schema.json",
+    "$schema": "http://json-schema.org/schema#",
+    "description": "Scratch 3.0 Project Schema",
+    "definitions": {
+        "optionalString": {
+            "oneOf": [
+                {"type": "string"},
+                {"type": "null"}
+            ]
+        },
+        "assetId": {
+            "type": "string",
+            "pattern": "^[a-fA-F0-9]{32}$"
+        },
+        "costume": {
+            "type": "object",
+            "properties": {
+                "assetId": { "$ref": "#/definitions/assetId"},
+                "bitmapResolution": {
+                    "type": "integer"
+                },
+                "dataFormat": {
+                    "type": "string",
+                    "enum": ["png", "PNG", "svg", "SVG", "jpeg", "JPEG", "jpg", "JPG", "bmp", "BMP"]
+                },
+                "name": {
+                    "type": "string"
+                },
+                "rotationCenterX": {
+                    "type": "number"
+                },
+                "rotationCenterY": {
+                    "type": "number"
+                },
+                "skinId": {
+                    "type": "integer"
+                }
+            },
+            "required": [
+                "assetId",
+                "dataFormat",
+                "name",
+                "rotationCenterX",
+                "rotationCenterY"
+            ]
+        },
+        "sound": {
+            "type": "object",
+            "properties": {
+                "assetId": { "$ref": "#/definitions/assetId"},
+                "dataFormat": {
+                    "type": "string",
+                    "enum": ["wav", "WAV", "wave", "WAVE", "mp3", "MP3"]
+                },
+                "format": {
+                    "type": "string",
+                    "enum": ["", "adpcm"]
+                },
+                "md5": {
+                    "type": "string",
+                    "pattern": "^[a-fA-F0-9]{32}.[a-zA-Z]+$"
```

Comment:
Actually, I realize that if we change the specification of the pattern later to remove the extension, older saved projects won't necessarily load any longer because pattern matching will fail. I think based on this new structure/specification being our goal, we'll actually have to make the VM change now rather than after save/load is released as a feature. The alternative is to allow either format in the sb3 validation, but deal with any discrepancies (e.g. removing the extension if it exists) in the vm...

One more alternative is to remove `md5` from the sb3_schema completely (or just have the constraint of it being a string without ever planning to add a pattern specification to it) and only have `assetId` and `dataFormat`. The VM is then free to do whatever it wants with `md5` in the future.

Reply:
I'm OK with changing (breaking) the SB3 format & spec as long as we do it before launch, or more specifically before we have "important" projects built on the spec. In my mind "important" projects are ones that are made by Scratchers or partners.

I'm also OK with having `assetId` and `dataFormat` as separate fields, or `md5` and `dataFormat`, or whatever you think makes the most sense.

The only thing I object to is having a field called "md5" which isn't actually (just) an MD5 -- mostly since we have other fields elsewhere in the code (not necessarily just this module) named "md5" which store simple MD5s and that gets confusing. :)

Reply:
Okay, after looking into what it would take to update everything (vm, gui, etc.) to actually reflect an md5 string where it says 'md5' and not an md5 string + a file extension, I have a new proposal.

I discovered that this will actually be a pretty pervasive change (I've already changed 19 files and I'm still not done apparently), since the name 'md5' is used to refer to an md5+ext all over the place, including the libraries we're loading. My proposal is to replace the property name `md5` with `md5ext` in this sb3 schema and the vm serializer/deserializers. That way, saved projects will only ever name the data correctly. For now, however, the serialization-deserialization code in the vm will translate md5ext back to the property name `md5` where necessary so that we don't break a million things in obscure places. How does that sound?

Reply:
That sounds good to me: "fixing" the serialized state seems higher priority so we're not saving files with the potentially-confusing field name, then we can be more thorough with the code & runtime state later.